spyder/api/plugins.py

```diff
@@ -206,213 +328,380 @@ def check_compatibility(self):
+        self._ismaximized = False
+        self._default_margins = None
+        self._isvisible = False
 
-    def switch_to_plugin(self):
-        """
-        Switch to this plugin.
+        self.shortcut = None
 
-        Notes
-        -----
-        This operation unmaximizes the current plugin (if any), raises
-        this plugin to view (if it's hidden) and gives it focus (if
-        possible).
-        """
-        super(BasePluginWidget, self)._switch_to_plugin()
+        # Options buttons
+        self.options_button = create_toolbutton(self, text=_('Options'),
+                                                icon=ima.icon('tooloptions'))
+        self.options_button.setPopupMode(QToolButton.InstantPopup)
 
+        # Don't show menu arrow and remove padding
+        if is_dark_interface():
+            self.options_button.setStyleSheet(
+                ("QToolButton::menu-indicator{image: none;}\n"
+                 "QToolButton{padding: 3px;}"))
+        else:
+            self.options_button.setStyleSheet(
+                "QToolButton::menu-indicator{image: none;}")
+
+        # Options menu
+        self._options_menu = QMenu(self)
+
+        # We decided to create our own toggle action instead of using
+        # the one that comes with dockwidget because it's not possible
+        # to raise and focus the plugin with it.
+        self._toggle_view_action = None
+
+        # Default actions for Options menu
+        self._dock_action = create_action(
+            self,
+            _("Dock"),
+            icon=ima.icon('dock'),
+            tip=_("Dock the pane"),
+            triggered=self._close_window,
+        )
+
+        self._undock_action = create_action(
+            self,
+            _("Undock"),
+            icon=ima.icon('undock'),
+            tip=_("Undock the pane"),
+            triggered=self._create_window,
+        )
+
+        self._close_plugin_action = create_action(
+            self,
+            _("Close"),
+            icon=ima.icon('close_pane'),
+            tip=_("Close the pane"),
+            triggered=self._plugin_closed,
+        )
+
+    # INFO: This is called from mainwindow.py
+    # This name does not convey that this can also (and should) be called
+    # from a non dockable widget. Non dockable widgets should be able to
+    # define a on_first_registration call
+    def _initialize_plugin_in_mainwindow_layout(self):
+        """
+        If this is the first time the plugin is shown, perform actions to
+        initialize plugin position in Spyder's window layout.
+
+        Use on_first_registration to define the actions to be run
+        by your plugin
+        """
+        if self.get_option('first_time', True):
+            try:
+                self.on_first_registration()
+            except NotImplementedError:
+                return
+            self.set_option('first_time', False)
+
+    # INFO: This is called from mainwindow.py
```

Comment:
What are these INFO comments?

Reply:
Things I left while I finish the PR. This is still not done.